### PR TITLE
TodoListView implemented for Todo app along with user authorisation check and data filtering

### LINKED ISSUE #2: Implement TodoListView

### PR DESCRIPTION
**Pull request by ankushchavan1**
Implemented features:
1. Created data model for Todo app.
2. Implemented TodoListView for the Todo app that will show the Todo list.
3. Implemented user authorisation check for the TodoListView that will check if the user is authorised before showing the data.
4. Implemented data filtering that will show only those Todo's to the user for which either user is the owner or the user is an editor.
5. Implemented abstract models for TimestampedModel and CreatorInfoModel.
6. Squashed migrations.

### CODE REVIEW

File: server/requirements/requirements.txt [hunk 1]
```diff
@@ -2,6 +2,8 @@ asgiref==3.4.1
 Django==3.2.6
 django-cors-headers==3.7.0
 django-environ==0.4.5
+django-model-utils==4.1.1
```
Comment (ronakmutha) [inline server/requirements/requirements.txt:5]: Why do you require this?

File: server/requirements/requirements.txt [hunk 2]
```diff
@@ -2,6 +2,8 @@ asgiref==3.4.1
...
+django-timestampedmodel==0.1.0
```
Comment (ronakmutha) [inline server/requirements/requirements.txt]: Why do you require these libraries?

File: server/apps/todo/models.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    title = models.CharField(max_length=100)
+    description = models.TextField(blank=True, null=True)
+    date = models.DateTimeField(default=now)
+    is_bookmarked = models.BooleanField(default=False)
+    editors = ArrayField(
+        models.CharField(max_length=100, blank=True),
+        size=5,
+        blank=True,
+        null=True,
+    )
```
Comment (ronakmutha) [inline server/apps/todo/models.py]: Editors are actual users, you can use Many to Many field here

File: server/apps/todo/migrations/0005_alter_todo_date.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        ('todo', '0004_auto_20210816_0523'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='todo',
+            name='date',
+            field=models.DateTimeField(default=django.utils.timezone.now),
+        ),
+    ]
```
Comment (ronakmutha) [inline server/apps/todo/migrations/0005_alter_todo_date.py]: Merge the migrations into one!

File: server/apps/todo/views.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        token = request.COOKIES.get('jwt')
+
+        if not token:
+            raise NotAuthenticated('Unauthenticated!')
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY, algorithms=['HS256'])
+        except jwt.ExpiredSignatureError:
+            raise NotAuthenticated('Authentication failed! Signature Expired!')
+
+        user = User.objects.filter(id=payload['id']).first()
```
Comment (ronakmutha) [inline server/apps/todo/views.py]: You can use `get` here, rather than first

File: server/apps/todo/models.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class Todo(TimestampedModel):
+    COMPLETED = "completed"
+    INPROGRESS = "inprogress"
+    STATUS = Choices((COMPLETED, COMPLETED), (INPROGRESS, INPROGRESS))
+    status = models.CharField(choices=STATUS, default=INPROGRESS, max_length=20)
+    owner = models.CharField(max_length=100)
+    created_by = models.CharField(max_length=100)
+    updated_by = models.CharField(max_length=100, blank=True, null=True)
```
Comment (ronakmutha) [inline server/apps/todo/models.py]: Write an abstract model. Check the pure drive code!

**Review by ronakmutha — CHANGES_REQUESTED** [review]

File: server/requirements/requirements.txt (hunk 2 above)
Comment (ankushchavan1) [reply server/requirements/requirements.txt]: I'm using `django-model-utils` to get `Choices` function that is used to set choices for status field in `todo/models.py`.

Using `django-timestampedmodel` to get TimestampedModel for implementing `Todo` model in `todo/models.py`. TimestampedModel will provide created_at and updated_at fields by default.

I just found another way to do this:
       class MyModel(models.Model):

            created_at = models.DateTimeField(auto_now_add=True)

            updated_at = models.DateTimeField(auto_now=True)

I will remove the `django-timestampedmodel` and implement the same using above snippet

File: server/apps/todo/migrations/0005_alter_todo_date.py (hunk 4 above)
Comment (ankushchavan1) [reply server/apps/todo/migrations/0005_alter_todo_date.py]: Squashed migrations in new commit

File: server/apps/todo/views.py (hunk 5 above)
Comment (ankushchavan1) [reply server/apps/todo/views.py]: Updated

File: server/apps/todo/models.py (hunk 3 above)
Comment (ankushchavan1) [reply server/apps/todo/models.py]: Updated with ManyToManyField type in latest commit

File: server/apps/todo/models.py (hunk 6 above)
Comment (ankushchavan1) [reply server/apps/todo/models.py]: Implemented abstract models in latest commit

File: server/requirements/requirements.txt (hunk 2 above)
Comment (ankushchavan1) [reply server/requirements/requirements.txt]: I have removed `django-timestampedmodel` dependecy and implemented the same behaviour by implementing `TimestampedModel` abstract model.

**Comment by ankushchavan1** [comment]
PR #4 is an extension of PR #3 . So, closing PR #3 unmerged.

**Pull request closed without merging**